File: static/js/hero-api.js
/**
 * Hero API Client
 * Handles communication with the Python backend for hero section data
 */

class HeroAPIClient {
    constructor() {
        this.baseURL = '/api/hero';
        this.cache = new Map();
        this.cacheTTL = 300000; // 5 minutes
    }

    /**
     * Make API request with error handling
     */
    async apiRequest(endpoint, options = {}) {
        const url = `${this.baseURL}${endpoint}`;
        const defaultOptions = {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json',
                'Accept': 'application/json'
            }
        };

        const requestOptions = { ...defaultOptions, ...options };

        try {
            const response = await fetch(url, requestOptions);
            
            if (!response.ok) {
                throw new Error(`API request failed: ${response.status} ${response.statusText}`);
            }

            const data = await response.json();
            
            if (!data.success) {
                throw new Error(data.error || data.message || 'API request failed');
            }

            return data.data;
        } catch (error) {
            console.error(`Hero API Error (${endpoint}):`, error);
            throw error;
        }
    }

    /**
     * Get cached data or fetch from API
     */
    async getCachedData(key, fetcher) {
        const cachedItem = this.cache.get(key);
        const now = Date.now();

        if (cachedItem && (now - cachedItem.timestamp) < this.cacheTTL) {
            return cachedItem.data;
        }

        try {
            const data = await fetcher();
            this.cache.set(key, {
                data,
                timestamp: now
            });
            return data;
        } catch (error) {
            // Return cached data if available, even if expired
            if (cachedItem) {
                console.warn('Using expired cache data due to API error');
                return cachedItem.data;
            }
            throw error;
        }
    }

    /**
     * Get complete hero section data
     */
    async getHeroData() {
        return this.getCachedData('hero_data', () => this.apiRequest('/'));
    }

    /**
     * Get hero statistics
     */
    async getHeroStats() {
        return this.getCachedData('hero_stats', () => this.apiRequest('/stats'));
    }

    /**
     * Get a specific hero statistic
     */
    async getHeroStat(statId) {
        return this.apiRequest(`/stats/${statId}`);
    }

    /**
     * Get typewriter text
     */
    async getTypewriterText(index = null) {
        const params = index !== null ? `?index=${index}` : '';
        return this.apiRequest(`/typewriter${params}`);
    }

    /**
     * Get next typewriter text
     */
    async getNextTypewriterText() {
        return this.apiRequest('/typewriter/next');
    }

    /**
     * Get profile information
     */
    async getProfileInfo() {
        return this.getCachedData('profile_info', () => this.apiRequest('/profile'));
    }

    /**
     * Get hero buttons
     */
    async getHeroButtons() {
        return this.getCachedData('hero_buttons', () => this.apiRequest('/buttons'));
    }

    /**
     * Increment a statistic
     */
    async incrementStat(statId) {
        return this.apiRequest(`/stats/${statId}/increment`, { method: 'POST' });
    }

    /**
     * Validate hero data
     */
    async validateHeroData() {
        return this.apiRequest('/validate');
    }

    /**
     * Get hero data as JSON
     */
    async getHeroJSON() {
        return this.apiRequest('/json');
    }

    /**
     * Health check
     */
    async healthCheck() {
        return this.apiRequest('/health');
    }

    /**
     * Clear cache
     */
    clearCache() {
        this.cache.clear();
    }

    /**
     * Initialize hero section with API data
     */
    async initializeHeroSection() {
        try {
            console.log('🚀 Initializing hero section with API data...');
            
            // Get hero data
            const heroData = await this.getHeroData();
            console.log('✅ Hero data loaded:', heroData);
            
            // Force hero elements visible first
            this.forceHeroElementsVisible();
            
            // Update statistics
            const stats = heroData.stats || await this.getHeroStats();
            this.updateHeroStats(stats);
            
            // Update typewriter
            if (heroData.typewriter_names) {
                this.initializeTypewriter(heroData.typewriter_names);
            }
            
            // Update buttons
            const buttons = heroData.buttons || await this.getHeroButtons();
            this.updateHeroButtons(buttons);
            
            console.log('✅ Hero section initialized successfully');
            return true;
            
        } catch (error) {
            console.error('❌ Failed to initialize hero section:', error);
            
            // Fallback: still try to make hero elements visible
            this.forceHeroElementsVisible();
            return false;
        }
    }

    /**
     * Update hero statistics in the DOM
     */
    updateHeroStats(stats) {
        try {
            const techStats = document.querySelector('.tech-stats');
            if (!techStats) {
                console.warn('Tech stats container not found');
                return;
            }

            stats.forEach(stat => {
                const statCard = document.querySelector(`[data-stat-id="${stat.id}"]`);
                if (statCard) {
                    const numberElement = statCard.querySelector('.stat-number');
                    const labelElement = statCard.querySelector('.stat-label');
                    const iconElement = statCard.querySelector('.stat-icon i');

                    if (numberElement) {
                        numberElement.textContent = stat.number + '+';
                        numberElement.setAttribute('data-target', stat.number);
                    }
                    if (labelElement) {
                        labelElement.textContent = stat.label;
                    }
                    if (iconElement) {
                        iconElement.className = stat.icon;
                    }
                }
            });

            console.log('✅ Hero stats updated');
        } catch (error) {
            console.error('❌ Failed to update hero stats:', error);
        }
    }

    /**
     * Update hero buttons in the DOM
     */
    updateHeroButtons(buttons) {
        try {
            const buttonsContainer = document.querySelector('.hero-buttons');
            if (!buttonsContainer) {
                console.warn('Hero buttons container not found');
                return;
            }

            buttons.forEach((button, index) => {
                const buttonElement = buttonsContainer.children[index];
                if (buttonElement) {
                    buttonElement.textContent = button.text;
                    buttonElement.href = button.href;
                    buttonElement.className = button.class;
                }
            });

            console.log('✅ Hero buttons updated');
        } catch (error) {
            console.error('❌ Failed to update hero buttons:', error);
        }
    }

    /**
     * Initialize typewriter with API data
     */
    initializeTypewriter(names) {
        try {
            if (window.typewriterNames) {
                window.typewriterNames = names;
                console.log('✅ Typewriter names updated');
            }
        } catch (error) {
            console.error('❌ Failed to initialize typewriter:', error);
        }
    }

    /**
     * Force refresh hero section stats
     */
    async refreshHeroStats() {
        try {
            this.clearCache();
            const stats = await this.getHeroStats();
            this.updateHeroStats(stats);
            
            // Trigger counter animations
            const techStats = document.querySelector('.tech-stats');
            if (techStats && window.animateCounters) {
                window.animateCounters(techStats);
            }
            
            console.log('✅ Hero stats refreshed');
        } catch (error) {
            console.error('❌ Failed to refresh hero stats:', error);
        }
    }

    /**
     * Force hero elements to be visible (delegated to animations.js)
     */
    forceHeroElementsVisible() {
        // Delegate to the main animation system to avoid conflicts
        if (window.forceHeroElementsVisible) {
            window.forceHeroElementsVisible();
        } else {
            console.warn('⚠️ Main animation system not loaded yet');
        }
    }
}

// Global instance
window.heroAPIClient = new HeroAPIClient();

// Auto-initialize when DOM is ready
document.addEventListener('DOMContentLoaded', async () => {
    console.log('🎯 DOM loaded, initializing hero section...');
    
    // Wait a moment for all elements to be in place
    setTimeout(async () => {
        try {
            await window.heroAPIClient.initializeHeroSection();
            
            // Additional fallback after 1 second
            setTimeout(() => {
                window.heroAPIClient.forceHeroElementsVisible();
            }, 1000);
            
        } catch (error) {
            console.error('Failed to auto-initialize hero section:', error);
            
            // Fallback: try to make elements visible anyway
            setTimeout(() => {
                window.heroAPIClient.forceHeroElementsVisible();
            }, 500);
        }
    }, 100);
});

// Export for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = HeroAPIClient;
}

File: hero_api.py
"""
Hero API Routes
API endpoints for hero section data and functionality
"""

from flask import Blueprint, jsonify, request
from hero_service import hero_service

# Create blueprint for hero API
hero_api = Blueprint('hero_api', __name__, url_prefix='/api/hero')

@hero_api.route('/', methods=['GET'])
def get_hero_data():
    """Get complete hero section data"""
    try:
        hero_data = hero_service.get_hero_data()
        return jsonify({
            "success": True,
            "data": hero_data,
            "message": "Hero data retrieved successfully"
        })
    except Exception as e:
        return jsonify({
            "success": False,
            "error": str(e),
            "message": "Failed to retrieve hero data"
        }), 500

@hero_api.route('/stats', methods=['GET'])
def get_hero_stats():
    """Get hero section statistics"""
    try:
        stats = hero_service.get_tech_stats()
        return jsonify({
            "success": True,
            "data": stats,
            "message": "Hero statistics retrieved successfully"
        })
    except Exception as e:
        return jsonify({
            "success": False,
            "error": str(e),
            "message": "Failed to retrieve hero statistics"
        }), 500

@hero_api.route('/stats/<stat_id>', methods=['GET'])
def get_hero_stat(stat_id):
    """Get a specific hero statistic"""
    try:
        stat = hero_service.get_stat_by_id(stat_id)
        if not stat:
            return jsonify({
                "success": False,
                "error": "Stat not found",
                "message": f"No statistic found with ID: {stat_id}"
            }), 404
        
        return jsonify({
            "success": True,
            "data": stat,
            "message": "Hero statistic retrieved successfully"
        })
    except Exception as e:
        return jsonify({
            "success": False,
            "error": str(e),
            "message": "Failed to retrieve hero statistic"
        }), 500

@hero_api.route('/typewriter', methods=['GET'])
def get_typewriter_text():
    """Get typewriter text"""
    try:
        index = request.args.get('index', type=int)
        text = hero_service.get_typewriter_text(index)
        return jsonify({
            "success": True,
            "data": {
                "text": text,
                "index": index,
                "all_texts": hero_service.typewriter_names
            },
            "message": "Typewriter text retrieved successfully"
        })
    except Exception as e:
        return jsonify({
            "success": False,
            "error": str(e),
            "message": "Failed to retrieve typewriter text"
        }), 500

@hero_api.route('/typewriter/next', methods=['GET'])
def get_next_typewriter_text():
    """Get next typewriter text in sequence"""
    try:
        text = hero_service.get_next_typewriter_text()
        return jsonify({
            "success": True,
            "data": {
                "text": text,
                "index": hero_service.current_name_index
            },
            "message": "Next typewriter text retrieved successfully"
        })
    except Exception as e:
        return jsonify({
            "success": False,
            "error": str(e),
            "message": "Failed to retrieve next typewriter text"
        }), 500

@hero_api.route('/profile', methods=['GET'])
def get_profile_info():
    """Get profile information"""
    try:
        profile = hero_service.get_profile_info()
        return jsonify({
            "success": True,
            "data": profile,
            "message": "Profile information retrieved successfully"
        })
    except Exception as e:
        return jsonify({
            "success": False,
            "error": str(e),
            "message": "Failed to retrieve profile information"
        }), 500

@hero_api.route('/buttons', methods=['GET'])
def get_hero_buttons():
    """Get hero section buttons"""
    try:
        buttons = hero_service.get_hero_buttons()
        return jsonify({
            "success": True,
            "data": buttons,
            "message": "Hero buttons retrieved successfully"
        })
    except Exception as e:
        return jsonify({
            "success": False,
            "error": str(e),
            "message": "Failed to retrieve hero buttons"
        }), 500

@hero_api.route('/stats/<stat_id>/increment', methods=['POST'])
def increment_stat(stat_id):
    """Increment a specific statistic"""
    try:
        result = hero_service.increment_stat(stat_id)
        if result:
            return jsonify({
                "success": True,
                "message": f"Statistic {stat_id} incremented successfully"
            })
        else:
            return jsonify({
                "success": False,
                "error": "Failed to increment statistic",
                "message": f"Could not increment statistic: {stat_id}"
            }), 400
    except Exception as e:
        return jsonify({
            "success": False,
            "error": str(e),
            "message": "Failed to increment statistic"
        }), 500

@hero_api.route('/validate', methods=['GET'])
def validate_hero_data():
    """Validate hero section data"""
    try:
        validation = hero_service.validate_hero_data()
        return jsonify({
            "success": True,
            "data": validation,
            "message": "Hero data validation completed"
        })
    except Exception as e:
        return jsonify({
            "success": False,
            "error": str(e),
            "message": "Failed to validate hero data"
        }), 500

@hero_api.route('/json', methods=['GET'])
def get_hero_json():
    """Get hero data as formatted JSON"""
    try:
        hero_json = hero_service.get_hero_json()
        return jsonify({
            "success": True,
            "data": hero_json,
            "message": "Hero JSON data retrieved successfully"
        })
    except Exception as e:
        return jsonify({
            "success": False,
            "error": str(e),
            "message": "Failed to retrieve hero JSON data"
        }), 500

# Health check endpoint
@hero_api.route('/health', methods=['GET'])
def health_check():
    """Health check for hero API"""
    return jsonify({
        "success": True,
        "data": {
            "status": "healthy",
            "service": "Hero API",
            "endpoints": [
                "GET /api/hero/",
                "GET /api/hero/stats",
                "GET /api/hero/stats/<stat_id>",
                "GET /api/hero/typewriter",
                "GET /api/hero/typewriter/next",
                "GET /api/hero/profile",
                "GET /api/hero/buttons",
                "POST /api/hero/stats/<stat_id>/increment",
                "GET /api/hero/validate",
                "GET /api/hero/json",
                "GET /api/hero/health"
            ]
        },
        "message": "Hero API is healthy"
    })
